lib/urlwatch/reporters.py

```diff
@@ -935,44 +938,129 @@ def _format_output(self, job_state):
         return summary_part, details_part
 
 
+if nio is not None:
+
+    class UrlwatchMatrixClient(nio.AsyncClient):
+        STORAGE_DIR = 'matrix_store'
+        CREDENTIALS_FILE = 'matrix-credentials.json'
+        DEFAULT_DEVICE_NAME = 'urlwatch'
+
+        def __init__(self, config, urlwatch_data_dir):
+            homeserver = config['homeserver']
+            user = config['user']
+            urlwatch_data_dir = Path(urlwatch_data_dir)
+            storage_dir = urlwatch_data_dir / self.STORAGE_DIR
+            storage_dir.mkdir(parents=True, exist_ok=True)
+            client_config = nio.AsyncClientConfig(max_limit_exceeded=0,
+                                                  max_timeouts=0,
+                                                  store_sync_tokens=True,
+                                                  encryption_enabled=True)
+            super().__init__(homeserver=homeserver,
+                             user=user,
+                             store_path=storage_dir,
+                             config=client_config)
+            self.urlwatch_data_dir = urlwatch_data_dir
+
+        async def login(self):
+            credentials_file = self.urlwatch_data_dir / self.CREDENTIALS_FILE
+            if credentials_file.is_file():
+                logger.debug("Reading stored credentials")
+                with credentials_file.open() as f:
+                    try:
+                        credentials = json.load(f)
+                        self.user_id = credentials['user_id']
+                        self.device_id = credentials['device_id']
+                        self.access_token = credentials['access_token']
+                    except json.JSONDecodeError:
+                        logger.warning(
+                            'Could not read cred json file, overwriting.')
+                self.load_store()
+
+            if not self.user_id or not self.device_id or not self.access_token:
+                logger.debug("Generating new credentials")
+                print("### Generating new Matrix credentials. ###")
+                device_name = input(f"Choose a name for this device: [{self.DEFAULT_DEVICE_NAME}] ")
+                if not device_name:
+                    device_name = self.DEFAULT_DEVICE_NAME
+                pw = getpass.getpass(f"Matrix password for {self.user} (it will not be stored): ")
```

Comment:
This should most likely be a separate mode from the command line, as we don't want random invocations to be interactive (urlwatch is most likely running as a cronjob or via some other non-interactive means).

Reply:
See the `--xmpp-login` command-line option for an example. Maybe combine it like `--test-filter` so that you can do e.g. `--login-reporter xmpp` and `--login-reporter matrix` (as `--xmpp-login` is an interactive thing, maybe just update docs + mention it in the list of deprecated things that `--xmpp-login is now `--login-reporter xmpp`).